### PR TITLE
rockchip: add support for nanopc t6

### PR DESCRIPTION
**Pull request by tian-xiaobo**
SoC: Rockchip RK3588
CPU: Quad-core ARM Cortex-A76(up to 2.4GHz) and quad-core Cortex-A55 CPU (up to 1.8GHz)
GPU: Mali-G610 MP4, compatible with OpenGLES 1.1, 2.0, and 3.2, OpenCL up to 2.2 and Vulkan1.2
VPU: 8K@60fps H.265 and VP9 decoder, 8K@30fps H.264 decoder, 4K@60fps AV1 decoder, 8K@30fps H.264 and H.265 encoder
NPU: 6TOPs, supports INT4/INT8/INT16/FP16
RAM: 64-bit 4GB/8GB/16GB LPDDR4X at 2133MHz
Flash: 32GB/64GB/256GB eMMC, at HS400 mode
microSD: support up to SDR104 mode
Ethernet: 2x PCIe 2.5G Ethernet
4G LTE: one mimiPCIe connector and one microSIM slot
USB-A: 1x USB 3.0 Type-A
USB-C: 1x Full function USB Type‑C™ port, support DP display up to 4Kp60, USB 3.0
Video input:
1x Standard HDMI input port, up to 4Kp60
2x 4-lane MIPI-CSI, compatible with MIPI V1.2
Video output:
2x Standard HDMI output ports
compatible with HDMI2.1, HDMI2.0, and HDMI1.4 operation
one support displays up to 7680x4320@60Hz, another one support up to 4Kp60
Support RGB/YUV(up to 10bit) format
2x 4-lane MIPI-DSI, compatible with MIPI DPHY 2.0 or CPHY 1.1
Audio:
3.5mm jack for stereo headphone output
2.0mm PH-2A connector for analog microphone input
GPIO:
40-pin 2.54mm header connector
up to 2x SPIs, 6x UARTs, 1x I2Cs, 8x PWMs, 2x I2Ss, 28x GPIOs
M.2 Connectors
one M.2 M-Key connector with PCIe 3.0 x4 for NVMe SSDs up to 2,500 MB/s
one M.2 E-key connector with PCIe 2.1 x1 and USB2.0 Host
others:
2 Pin 1.27/1.25mm RTC battery input connector for low power RTC IC HYM8563TS
one 38Khz IR receiver
MASK button for eMMC update, reset button, and Power button
one 5V Fan connector
Debug UART, 3-Pin 2.54mm header, 3.3V level, 1500000bps
2 x GPIO Controlled LED (SYS, LED1)
Power supply: 5.5*2.1mm DC Jack, 12VDC input.
PCB: 8 Layer, 110x80x1.6mm
Ambient Operating Temperature: 0℃ to 70℃

### CODE REVIEW

**Review by peterwillcn — APPROVED** [review]

File: target/linux/rockchip/armv8/config-6.6
Comment (hauke) [inline target/linux/rockchip/armv8/config-6.6:1]: If your board needs special driver please use kernel modules and do not modify the main kernel configuration if possible.

Comment (peterwillcn) [reply target/linux/rockchip/armv8/config-6.6:1]: fixed

File: package/boot/arm-trusted-firmware-rockchip/Makefile [hunk 1]
```diff
@@ -7,10 +7,10 @@
 
 include $(TOPDIR)/rules.mk
 
-PKG_VERSION:=2.10
+PKG_VERSION:=2.10.2
```
Comment (1715173329) [inline package/boot/arm-trusted-firmware-rockchip/Makefile:10]: unrelated change, this should be done in a separated PR.

File: package/boot/uboot-rockchip/Makefile [hunk 2]
```diff
@@ -229,6 +229,13 @@ define U-Boot/nanopi-r6s-rk3588s
     friendlyarm_nanopi-r6s
 endef
 
+define U-Boot/nanopc-t6-rk3588
```
Comment (1715173329) [inline package/boot/uboot-rockchip/Makefile:232]: this should put above nanopi-r6s-rk3588s

File: target/linux/rockchip/armv8/base-files/etc/board.d/01_leds [hunk 3]
```diff
@@ -18,6 +18,7 @@ xunlong,orangepi-r1-plus-lts)
 	ucidef_set_led_netdev "wan" "WAN" "green:wan" "eth0"
 	ucidef_set_led_netdev "lan" "LAN" "green:lan" "eth1"
 	;;
+friendlyarm,nanopc-t6|\
```
Comment (1715173329) [inline target/linux/rockchip/armv8/base-files/etc/board.d/01_leds:21]: T6 does not have any of these LEDs.

File: target/linux/rockchip/armv8/base-files/etc/board.d/02_network [hunk 4]
```diff
@@ -17,6 +17,7 @@ rockchip_setup_interfaces()
 	xunlong,orangepi-r1-plus-lts)
 		ucidef_set_interfaces_lan_wan 'eth1' 'eth0'
 		;;
+	friendlyarm,nanopc-t6|\
```
Comment (1715173329) [inline target/linux/rockchip/armv8/base-files/etc/board.d/02_network:20]: please put this section above `friendlyarm,nanopi-r2c`

File: target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity [hunk 5]
```diff
@@ -43,6 +43,7 @@ friendlyarm,nanopi-r4s-enterprise)
 	set_interface_core 10 "eth0"
 	set_interface_core 20 "eth1"
 	;;
+friendlyarm,nanopc-t6|\
```
Comment (1715173329) [inline target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity:46]: ditto

File: target/linux/rockchip/image/armv8.mk [hunk 6]
```diff
@@ -89,6 +89,14 @@ define Device/friendlyarm_nanopi-r6s
 endef
 TARGET_DEVICES += friendlyarm_nanopi-r6s
 
+define Device/friendlyarm_nanopc-t6
```
Comment (1715173329) [inline target/linux/rockchip/image/armv8.mk:92]: ditto

**Review by 1715173329 — CHANGES_REQUESTED** [review]

File: package/boot/arm-trusted-firmware-rockchip/Makefile (hunk 1 above)
Comment (peterwillcn) [reply package/boot/arm-trusted-firmware-rockchip/Makefile]: fixed

File: package/boot/uboot-rockchip/Makefile (hunk 2 above)
Comment (peterwillcn) [reply package/boot/uboot-rockchip/Makefile]: fixed

File: target/linux/rockchip/armv8/base-files/etc/board.d/01_leds (hunk 3 above)
Comment (peterwillcn) [reply target/linux/rockchip/armv8/base-files/etc/board.d/01_leds]: removed

File: target/linux/rockchip/armv8/base-files/etc/board.d/02_network (hunk 4 above)
Comment (peterwillcn) [reply target/linux/rockchip/armv8/base-files/etc/board.d/02_network]: fixed

File: target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity (hunk 5 above)
Comment (peterwillcn) [reply target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity:46]: r2c usb eth1

File: target/linux/rockchip/image/armv8.mk (hunk 6 above)
Comment (peterwillcn) [reply target/linux/rockchip/image/armv8.mk:84]: fixed

File: target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity (hunk 5 above)
Comment (1715173329) [reply target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity:46]: the whole section should be put above `friendlyarm,nanopi-r2c`

File: package/boot/uboot-rockchip/Makefile [hunk 7]
```diff
@@ -222,6 +222,13 @@ define U-Boot/Default/rk3588
   TPL:=rk3588_ddr_lp4_2112MHz_lp5_2400MHz_v1.16.bin
 endef
 
+define U-Boot/nanopc-t6-rk3588
+  $(U-Boot/Default/rk3588)
+  NAME:=NanoPc T6
```
Comment (1715173329) [inline package/boot/uboot-rockchip/Makefile:227]:
```suggestion
  NAME:=NanoPC T6
```

from the [documentaion](https://wiki.friendlyelec.com/wiki/index.php/NanoPC-T6)

File: target/linux/rockchip/image/armv8.mk [hunk 8]
```diff
@@ -22,6 +22,14 @@ define Device/friendlyarm_nanopc-t4
 endef
 TARGET_DEVICES += friendlyarm_nanopc-t4
 
+define Device/friendlyarm_nanopc-t6
+  DEVICE_VENDOR := FriendlyARM
+  DEVICE_MODEL := NanoPc T6
```
Comment (1715173329) [inline target/linux/rockchip/image/armv8.mk:27]:
```suggestion
  DEVICE_MODEL := NanoPC T6
```

File: target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity [hunk 9]
```diff
@@ -29,6 +29,7 @@ set_interface_core() {
 }
 
 case "$(board_name)" in
+friendlyarm,nanopc-t6|\
```
Comment (1715173329) [inline target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity:32]: sorry if I made any confusing, it should be like this:
```
friendlyarm,nanopc-t6|\
friendlyarm,nanopi-r5c|\
radxa,e25|\
sinovoip,rk3568-bpi-r2pro)
	set_interface_core 2 "eth0"
	set_interface_core 4 "eth1"
	;;
friendlyarm,nanopi-r2c|\
friendlyarm,nanopi-r2c-plus|\
friendlyarm,nanopi-r2s|\
radxa,cm3-io|\
xunlong,orangepi-r1-plus|\
xunlong,orangepi-r1-plus-lts)
	set_interface_core 2 "eth0"
	set_interface_core 4 "eth1" "xhci-hcd:usb[0-9]+"
	;;
[...]
```

and the network part:
```
	friendlyarm,nanopc-t6|\
	friendlyarm,nanopi-r5c|\
	radxa,e25)
		ucidef_set_interfaces_lan_wan 'eth0' 'eth1'
		;;
	friendlyarm,nanopi-r2c|\
	friendlyarm,nanopi-r2c-plus|\
	friendlyarm,nanopi-r2s|\
	friendlyarm,nanopi-r4s|\
	friendlyarm,nanopi-r4s-enterprise|\
	radxa,rockpi-e|\
	xunlong,orangepi-r1-plus|\
	xunlong,orangepi-r1-plus-lts)
		ucidef_set_interfaces_lan_wan 'eth1' 'eth0'
		;;
[...]
```

**Comment by 1715173329** [comment]
please also create a patch for adding OpenWrt LED aliases like https://github.com/openwrt/openwrt/blob/main/target/linux/rockchip/patches-6.6/115-rock-3a-add-led-aliases-and-stop-heartbeat.patch.

File: target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity (hunk 9 above)
Comment (tian-xiaobo) [reply target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity:32]: Thank you for the example code.

**Comment by tian-xiaobo** [comment]
> please also create a patch for adding OpenWrt LED aliases like https://github.com/openwrt/openwrt/blob/main/target/linux/rockchip/patches-6.6/115-rock-3a-add-led-aliases-and-stop-heartbeat.patch. and mmc speed https://github.com/openwrt/openwrt/blob/c365bd109cdd01c00da8910be568d739fbcbcc73/target/linux/rockchip/patches-6.6/119-arm64-dts-rockchip-lower-mmc-speed-for-Radxa-Rock-5B.patch

Done

**Comment by tian-xiaobo** [comment]
> I mean do the same thing for NanoPC T6 ;)

Sorry, I misunderstood. I'll study it :-)

**Comment by tian-xiaobo** [comment]
> I mean do the same thing for NanoPC T6 ;)

Heartbeat LED is the default behavior, and I think there is no need for special customization or modification.
[Fixed lower mmc speed for friendlyarm nanopc t6](https://github.com/openwrt/openwrt/pull/16158/files#diff-b21dedcea6767f3e56933446994b8f217875d8d6bc1d9601d467c0862c129726)

**Comment by 1715173329** [comment]
> Heartbeat LED is the default behavior

Actually, disable heartbeat and add aliases. We use led-{boot,failsafe,running,upgrade} bindings to control the LED.

**Comment by tian-xiaobo** [comment]
> > Heartbeat LED is the default behavior
> 
> Actually, disable heartbeat and add aliases. We use led-{boot,failsafe,running,upgrade} bindings to control the LED.

Oh, I see. I understand, I've learned. I'll make some modifications

**Comment by tian-xiaobo** [comment]
> > Heartbeat LED is the default behavior
> 
> Actually, disable heartbeat and add aliases. We use led-{boot,failsafe,running,upgrade} bindings to control the LED.

Done

**Comment by tian-xiaobo** [comment]
> > Heartbeat LED is the default behavior
> 
> Actually, disable heartbeat and add aliases. We use led-{boot,failsafe,running,upgrade} bindings to control the LED.

DONE

**Comment by tian-xiaobo** [comment]
fix patch error

File: target/linux/rockchip/patches-6.6/036-v6.11-arm64-dts-rockchip-lower-mmc-speed-for-nanopc-t6.patch [hunk 10]
```diff
@@ -0,0 +1,26 @@
+From 564de3b973e0851dc06127a2e22894d7cdf58a2e Mon Sep 17 00:00:00 2001
+From: xiaobo <peterwillcn@gmail.com>
+Date: Wed, 14 Aug 2024 18:46:28 +0800
+Subject: [PATCH] arm64:dts:rockchip: lower mmc speed for friendlyarm nanopc t6
```
Comment (1715173329) [inline target/linux/rockchip/patches-6.6/036-v6.11-arm64-dts-rockchip-lower-mmc-speed-for-nanopc-t6.patch:4]:
```suggestion
Subject: [PATCH] arm64: dts: rockchip: lower mmc speed for friendlyarm nanopc t6
```

File: target/linux/rockchip/patches-6.6/037-v6.11-arm64-dts-rockchip-add-led-aliases-and-stop-heartbea-for-nanopc-t6.patch [hunk 11]
```diff
@@ -0,0 +1,38 @@
+From 8bcc1b79e465e8e0f00e99625cf848d7b3fdd764 Mon Sep 17 00:00:00 2001
+From: xiaobo <peterwillcn@gmail.com>
+Date: Wed, 14 Aug 2024 20:46:38 +0800
+Subject: [PATCH] arm64:dts:rockchip: add led aliases and stop heartbeat for
```
Comment (1715173329) [inline target/linux/rockchip/patches-6.6/037-v6.11-arm64-dts-rockchip-add-led-aliases-and-stop-heartbea-for-nanopc-t6.patch:4]:
```suggestion
Subject: [PATCH] arm64: dts: rockchip: add led aliases and stop heartbeat for
```

**Review by 1715173329 — COMMENTED** [review]
refresh kernel patches with `make target/linux/{clean,refresh}` V=s

**Comment by tian-xiaobo** [comment]
> refresh kernel patches with `make target/linux/{clean,refresh}` V=s

I found an issue where 6.6.45 in OpenWrt does not match the hash in the Linux code repository~
image

**Comment by tian-xiaobo** [comment]
> refresh kernel patches with `make target/linux/{clean,refresh}` V=s

Sorry, I found the problem and it conflicts with 032-032-v6.8-arm64 dts-rockchip-move rk3588 serial aliases to soc-dtsi patch.

**Comment by 1715173329** [comment]
The kernel patches still need refresh.

**Comment by peterwillcn** [comment]
> The kernel patches still need refresh.

Can you provide a tutorial on target/Linux/refresh?

**Comment by peterwillcn** [comment]
> The kernel patches still need refresh.

refresh done

**Review by peterwillcn — APPROVED** [review]

**Comment by tian-xiaobo** [comment]
ping @hauke

**Comment by peterwillcn** [comment]
ping @1715173329
